### PR TITLE
Add managed .NET gRPC client to interop tests

### PR DESCRIPTION
**Pull request by JamesNK**
Addresses https://github.com/grpc/grpc-dotnet/issues/209

Blocked on https://github.com/grpc/grpc-dotnet/pull/212 being merged

### CODE REVIEW

**Comment by JamesNK** [comment]
The managed gRPC client + interop client have been merged

This is ready to be tested and merged

**Comment by JamesNK** [comment]
btw these are the tests that pass at the moment:

*  "empty_unary"
*  "large_unary"
*  "client_streaming"
*  "server_streaming"
*  "empty_stream"
*  "cancel_after_begin"
*  "timeout_on_sleeping_server"
*  "custom_metadata"
*  "status_code_and_message"
*  "unimplemented_service"
*  "unimplemented_method"

OAuth is not done, client compression is not done, and there is a limitation in HttpClient that prevents duplex streaming that causes `ping_pong` and `cancel_after_first_response` to hang. You'll need to exclude them.

**Comment by jtattermusch** [comment]
I pushed a few run_interop_tests.py fixes, let's see what the tests say.

**Comment by JamesNK** [comment]
The interop test client wasn't using TLS correctly when it was enabled. Fixed here in grpc-dotnet - https://github.com/grpc/grpc-dotnet/pull/230

File: tools/run_tests/run_interop_tests.py [hunk 1]
```diff
@@ -210,8 +209,10 @@ def global_env(self):
         return {}
 
     def unimplemented_test_cases(self):
-        # aspnetcore doesn't have a client so ignore all test cases.
-        return _TEST_CASES + _AUTH_TEST_CASES
+        return _SKIP_COMPRESSION + \
+            _SKIP_SPECIAL_STATUS_MESSAGE + \
+            _AUTH_TEST_CASES + \
+            ['cancel_after_first_response', 'ping_pong']
```
Comment (JamesNK) [inline tools/run_tests/run_interop_tests.py]: This is temporary. When HttpClient streaming has been fixed these will be enabled.

**Comment by JamesNK** [comment]
https://source.cloud.google.com/results/invocations/b97f0bdb-5797-4f24-b9b1-54b30875a5b7/targets/github%2Fgrpc%2Finterop_test/tests;group=tests;test=cloud_to_cloud:aspnetcore:java_server:unimplemented_service:tls;row=39

`System.InvalidOperationException: Response did not have a grpc-status trailer.`

In the new gRPC client I am checking that a status was returned and throwing an error if it is not available. Is returning a grpc-status trailer optional? It seems like the Java, Dart and Go servers are skipping it (or there is a bug in new client that is stopping it from reading the status)

**Comment by jtattermusch** [comment]
> https://source.cloud.google.com/results/invocations/b97f0bdb-5797-4f24-b9b1-54b30875a5b7/targets/github%2Fgrpc%2Finterop_test/tests;group=tests;test=cloud_to_cloud:aspnetcore:java_server:unimplemented_service:tls;row=39
> 
> `System.InvalidOperationException: Response did not have a grpc-status trailer.`
> 
> In the new gRPC client I am checking that a status was returned and throwing an error if it is not available. Is returning a grpc-status trailer optional? It seems like the Java, Dart and Go servers are skipping it (or there is a bug in new client that is stopping it from reading the status)
> 
> Edit: I'm pretty sure this is an error in either HttpClient or the .NET Core gRPC client. Will investigate more.

I'd be very very surprised if grpc-status was optional. Seems more like a bug in the http client. Btw, is this happening for any other test than `unimplemented_service`?

**Comment by JamesNK** [comment]
@jtattermusch Why are the interop tests not running anymore? Did I do something in this PR or is it an environmental problem?

**Comment by jtattermusch** [comment]
@JamesNK  the "build_docker_nodepurejs

> @jtattermusch Why are the interop tests not running anymore? Did I do something in this PR or is it an environmental problem?

The failure is not related to your code. It looks like the build_docker_nodepurejs has failed (probably some transient error with Node.js) so the entire build has failed and the tests haven't run. I'll restart the tests
https://source.cloud.google.com/results/invocations/5c9cb5bc-54b4-4881-8dd7-a7943ab1b402/targets/github%2Fgrpc%2Finterop_docker_build/tests

**Comment by jtattermusch** [comment]
@JamesNK  from your experience, are the managed client interop tests passing reliably?  I'd like to prevent introducing interop tests flakes by merging this PR.

**Comment by jtattermusch** [comment]
I still see some failures with the managed client:
https://source.cloud.google.com/results/invocations/cb7102ea-bd3b-4e26-971d-b9fce8e2e236/targets/github%2Fgrpc%2Finterop_test/tests;group=tests;test=cloud_to_cloud:nodepurejs:java_server:server_streaming:tls;row=33

**Comment by JamesNK** [comment]
Need to merge https://github.com/grpc/grpc-dotnet/pull/240 before they all pass

**Comment by JamesNK** [comment]
@jtattermusch All the enabled managed interop tests are passing. This can be merged 🎉

**Comment by jtattermusch** [comment]
Known failures:
https://github.com/grpc/grpc/issues/18967
https://github.com/grpc/grpc/issues/18968

**Pull request merged**